### DISCUSSION TITLE
Computed state which relies on current time

### DISCUSSION THREAD

**Discussion by ninjz** [pmndrs/zustand#2150] [Q&A]
I have a computed state which I want to be re-computed when its dependent states are updated AND when the current time exceeds the date interval returned from another computed state.

I've come up with this so far:
```js
export const useCurrentState = () => {
    const now = new Date();

    const currentInterval = useStore(state =>
        state.getInterval(now),
    );
    const currentState = useStore(state =>
        state.getSchedulerState(now),
    );

    const [tick, setTick] = useState(0);

    useEffect(() => {
        const now = new Date();
        // Calculate the remaining time in the current interval
        const remainingTime = differenceInMilliseconds(currentInterval, now);

        // Set a timer to force a re-render when the current interval ends
        const timer = setTimeout(() => {
            setTick(tick + 1);
        }, remainingTime);
        
        // Clean up the timer when the component unmounts or the state changes
        return () => clearTimeout(timer);
    }, [currentInterval, tick]);

    return currentState;
};
```

I would like to be able to accomplish this without having to create a separate hook on top of the store. I've also considered having the current time being a state in the store and updating it every second however I'm concerned about the performance impact of that.

What is the best way to achieve this?

**Comment by dai-shi** [comment]
Using `useEffect` is idiomatic with React, because you can cleanly define the cleanup function and let React handle it with their lifecycle.

If you want to do without React, you could do it like so.
```js
const useFooStore = create(() => ({ now: 0 })
setInterval(() => {
  useFooStore.setState({ now: Date.now() })
}, 1000)
```
That's just a naive example, not directly related with your use requirement. But, I hope you get the idea.

**Reply by ninjz** [reply]
Thank you for your response. Do you see any issues with updating the store while recomputing other states every second in a react native mobile app? (My functions are pure and solely rely on the store's state so I feel it shouldn't be computationally expensive.)

I need this to pretty much be running throughout the entire life cycle of the app. 

Would you suggest setting this outside of React (like with 
your example) or run this in a suitable component using something like my hook example?

**Reply by dai-shi** [reply]
It depends on your app requirement. Running tasks every second can be trivial, but it may consume batteries right?

Doing it either outside React or inside is valid, depending on your architecture.

**Comment by venkatd** [comment]
@ninjz What are you trying to accomplish in the interface? There might be simpler ways to accomplish depending on what your goals are. Having the zustand store update its state every second sounds quit complex and resource intensive.
